engine/equivalence.ml

```diff
@@ -0,0 +1,85 @@
+open Merge_accessors
+
+type specialized_tree =
+  | Terminal of constraint_tree
+  | SNode of sym_value * (node_constraint * specialized_tree) list * (node_constraint * specialized_tree) option
+and
+  node_constraint =
+  | Symbolic of domain
+  | Resolved of bool
+
+let compare (src: Merge_accessors.constraint_tree) (dst: Merge_accessors.constraint_tree) : bool =
+  let rec init : Merge_accessors.constraint_tree -> specialized_tree = function
+    | (Leaf _ | Failure) as t -> Terminal t
+    | Node (acc, children, fallback) ->
+      let c = children |> List.map (fun (pi, c_tree) -> Symbolic pi, init c_tree) in
+      let fb = fallback |> Option.map (fun (pi, c_tree) -> Symbolic pi, init c_tree) in
+      SNode (acc, c, fb)
+  in
+  let specialize (pi: domain) (pi': domain) : node_constraint =
+    let intersect = Target_sym_engine.Domain.inter in
+    let is_empty (set:domain) =
+      Target_sym_engine.Domain.Set.is_empty set.int && Target_sym_engine.Domain.Set.is_empty set.tag
+    in
+    match intersect pi pi' with
+    | x when x = pi' ->
+      Resolved true
```

Comment:
There is a bug lurking here, which is the assumption that the language default equality `(=)` coincides with the notion of equality we want on `Diet.t` value. Depending on how `Diet.t` is implemented, this may not be true (for example, if sets of disjoint intervals are represented as balanced trees of stuff, then different balancing choices may lead to physically-distinct representations for the same sets). One should instead use a function `Target_sym_engine.Domain.equal`.

Reply:
It is not clear to me why the equality is on `pi'` and not `pi`.

If `src` is the constraint that comes from the *other* tree, and `dst` is the constraint on the current node we are looking at, the criterion to resolve to `true` would be, I think, that `src` is included in `dst`: the only cases that we are considering all respect `dst` (the local constraint), so our node with this constraint will always be selected. To check that `src` is included in `dst`, I would check `Domain.equal src (Domain.inter src dst)`, not the other way around.

For a concrete example: if the source constraint (pi?) is [2;3] and the local constraint [1;5] (pi'?), then I think the local child will always be chosen so we should get `Resolve true`, but `inter [2;3] [1;5]` is `[2;3]`, so (if I understood which variable means which here) we have `Domain.equal x pi` and not `Domain.equal x pi'`.